Ingress and use TotalReadoutTime from epi fmap meta-data in FSL topup

sgiavasis opened the pull request:
<!-- If you'd like use an emoji in the PR title, consider checking https://gitmoji.dev for guidance on emoji selection. Clicking an emoji image on that page will copy it to your clipboard. -->
## Fixes
<!-- If PR doesn't fully resolve the issue, replace 'Fixes' below with 'Related to'. -->
<!-- If there is no issue being resolved, open one before creating this pull request. -->
<!-- The square brackets in the template represent something for you to replace. For example, you might change "Fixes #[issue number]" to "Fixes #1". -->
Fixes #1768 by @sgiavasis 

## Description
- CPAC now reads in `TotalReadoutTime` from `_epi` field map meta-data, from the JSON sidecars.
- If `Blip-FSL-TOPUP` is enabled, and `TotalReadoutTime` is present, it will use this first. If not, it will search for and use `DwellTime` from the meta-data instead (the current default behavior).

## Tests
<!-- Give steps for the reviewer to verify that this PR fixes the problem; or delete the section entirely. -->
- [x] When `TotalReadoutTime` and `DwellTime` both present in meta-data, use `TotalReadoutTime`.
- [x] Works with only `TotalReadoutTime`
- [x] Works with only `DwellTime`
- [x] Clear message when both are missing

## Checklist
<!-- Replace  the [ ] with [x] to check the boxes. -->
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [x] My pull request targets the `develop` branch of the repository. <!-- Change this branch if you're targeting a branch other than `develop` -->
- [x] My commit messages follow best practices.
- [x] My code follows the established code style of the repository.
- [ ] I added tests for the changes I made (if applicable).
- [x] I updated the changelog.
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.

## Developer Certificate of Origin
<!-- You must read and understand the following attestation. -->

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

shnizzedy commented:
I'm still getting crashes like

```Python
C-PAC failed to start
Traceback (most recent call last):
  File "/code/CPAC/pipeline/cpac_runner.py", line 577, in run
    p_name, plugin, plugin_args, test_config)
  File "/code/CPAC/pipeline/cpac_pipeline.py", line 448, in run_workflow
    raise exception
  File "/code/CPAC/pipeline/cpac_pipeline.py", line 444, in run_workflow
    subject_id, sub_dict, c, p_name, num_ants_cores
  File "/code/CPAC/pipeline/cpac_pipeline.py", line 1117, in build_workflow
    wf, rpool = initiate_rpool(wf, cfg, sub_dict)
  File "/code/CPAC/pipeline/engine.py", line 1948, in initiate_rpool
    part_id, ses_id)
  File "/code/CPAC/pipeline/engine.py", line 1526, in ingress_raw_func_data
    data_paths['creds_path'], ses_id)
  File "/code/CPAC/utils/datasource.py", line 487, in ingress_func_metadata
    node, out_file = rpool.get('diffphase-dwell')[
  File "/code/CPAC/pipeline/engine.py", line 276, in get
    raise LookupError(info_msg)
LookupError: 

[!] C-PAC says: None of the listed resources are in the resource pool:

  diffphase-dwell

Options:
- You can enable a node block earlier in the pipeline which produces these resources. Check the 'outputs:' field in a node block's documentation.
- You can directly provide this required data by pulling it from another BIDS directory using 'source_outputs_dir:' in the pipeline configuration, or by placing it directly in your C-PAC output directory.
- If you have done these, and you still get this message, please let us know through any of our support channels at: https://fcp-indi.github.io/
```

with https://github.com/FCP-INDI/C-PAC/blob/99bdfb39a45816155935a89e924ab4d12ab41f0b/CPAC/resources/configs/pipeline_config_rbc-options.yml#L101-L104 and data like 
```tree
.
├── anat
│   ├── sub-2207352563_ses-PNC1_acq-refaced_T1w.json
│   └── sub-2207352563_ses-PNC1_acq-refaced_T1w.nii.gz
├── fmap
│   ├── sub-2207352563_ses-PNC1_magnitude1.json
│   ├── sub-2207352563_ses-PNC1_magnitude1.nii.gz
│   ├── sub-2207352563_ses-PNC1_magnitude2.json
│   ├── sub-2207352563_ses-PNC1_magnitude2.nii.gz
│   ├── sub-2207352563_ses-PNC1_phase1.json
│   ├── sub-2207352563_ses-PNC1_phase1.nii.gz
│   ├── sub-2207352563_ses-PNC1_phase2.json
│   └── sub-2207352563_ses-PNC1_phase2.nii.gz
└── func
    ├── sub-2207352563_ses-PNC1_task-frac2back_bold.json
    ├── sub-2207352563_ses-PNC1_task-frac2back_bold.nii.gz
    ├── sub-2207352563_ses-PNC1_task-frac2back_events.tsv
    ├── sub-2207352563_ses-PNC1_task-idemo_acq-VARIANTNumVolumes_bold.json
    ├── sub-2207352563_ses-PNC1_task-idemo_acq-VARIANTNumVolumes_bold.nii.gz
    └── sub-2207352563_ses-PNC1_task-idemo_acq-VARIANTNumVolumes_events.tsv
```
And these sidecars (ones with `TotalReadoutTime` pasted here, none have `DwellTime`, [full set](https://github.com/FCP-INDI/C-PAC/files/9558662/sub-2207352563_ses-PNC1_sidecars.zip) attached):

#### `sub-2207352563_ses-PNC1_task-frac2back_bold.json`
```JSON
{
    "AcquisitionMatrixPE": 64,
    "AcquisitionNumber": 1,
    "BandwidthPerPixelPhaseEncode": 28.409,
    "BaseResolution": 64,
    "ConversionSoftware": "dcm2niix",
    "ConversionSoftwareVersion": "v1.0.20181114  (JP2:OpenJPEG) GCC4.8.4",
    "DeidentificationMethod": "Penn_BSC_profile_v1.0",
    "DerivedVendorReportedEchoSpacing": 0.000550002,
    "EchoTime": 0.032,
    "EffectiveEchoSpacing": 0.000550002,
    "FlipAngle": 90,
    "ImageOrientationPatientDICOM": [
        1,
        0,
        0,
        0,
        1,
        0
    ],
    "ImageType": [
        "ORIGINAL",
        "PRIMARY",
        "M",
        "ND",
        "MOSAIC"
    ],
    "ImagingFrequency": 123.259,
    "InPlanePhaseEncodingDirectionDICOM": "COL",
    "MRAcquisitionType": "2D",
    "MagneticFieldStrength": 3,
    "Manufacturer": "Siemens",
    "ManufacturersModelName": "TrioTim",
    "Modality": "MR",
    "PartialFourier": 1,
    "PatientPosition": "HFS",
    "PercentPhaseFOV": 100,
    "PhaseEncodingDirection": "j-",
    "PhaseEncodingSteps": 64,
    "PhaseResolution": 1,
    "PixelBandwidth": 2056,
    "ProcedureStepDescription": "mri_brain",
    "ProtocolName": "bbl1_frac2back1_231",
    "PulseSequenceDetails": "%SiemensSeq%_ep2d_bold",
    "ReceiveCoilName": "32Ch_Head",
    "ReconMatrixPE": 64,
    "RepetitionTime": 3,
    "SAR": 0.253013,
    "ScanOptions": "FS",
    "ScanningSequence": "EP",
    "SequenceName": "_epfid2d1_64",
    "SequenceVariant": "SK",
    "SeriesDescription": "bbl1_frac2back1_231",
    "SeriesNumber": 8,
    "ShimSetting": [
        4067,
        -6772,
        -3778,
        1390,
        -75,
        -153,
        624,
        34
    ],
    "SliceThickness": 3,
    "SliceTiming": [
        1.505,
        0,
        1.57,
        0.065,
        1.6375,
        0.13,
        1.7025,
        0.195,
        1.7675,
        0.2625,
        1.8325,
        0.3275,
        1.8975,
        0.3925,
        1.965,
        0.4575,
        2.03,
        0.5225,
        2.095,
        0.59,
        2.16,
        0.655,
        2.225,
        0.72,
        2.29,
        0.785,
        2.3575,
        0.85,
        2.4225,
        0.9175,
        2.4875,
        0.9825,
        2.5525,
        1.0475,
        2.6175,
        1.1125,
        2.685,
        1.1775,
        2.75,
        1.2425,
        2.815,
        1.31,
        2.88,
        1.375,
        2.945,
        1.44
    ],
    "SoftwareVersions": "syngo_MR_B17",
    "SpacingBetweenSlices": 3,
    "StudyID": "7686972",
    "StudyInstanceUID": "1.2.840.113745.101000.1002000.40728.4508.30537163",
    "TaskName": "frac2back",
    "TotalReadoutTime": 0.0346501,
    "TxRefAmp": 330.354,
    "Obliquity": "False",
    "VoxelSizeDim1": 3.0,
    "VoxelSizeDim3": 3.0,
    "VoxelSizeDim2": 3.0,
    "Dim1Size": 64,
    "Dim2Size": 64,
    "Dim3Size": 46,
    "NumVolumes": 231
}
```

#### `sub-2207352563_ses-PNC1_task-idemo_acq-VARIANTNumVolumes_bold.json`
```JSON
{
    "AcquisitionMatrixPE": 64,
    "AcquisitionNumber": 1,
    "BandwidthPerPixelPhaseEncode": 28.409,
    "BaseResolution": 64,
    "ConversionSoftware": "dcm2niix",
    "ConversionSoftwareVersion": "v1.0.20181114  (JP2:OpenJPEG) GCC4.8.4",
    "DeidentificationMethod": "Penn_BSC_profile_v1.0",
    "DerivedVendorReportedEchoSpacing": 0.000550002,
    "EchoTime": 0.032,
    "EffectiveEchoSpacing": 0.000550002,
    "FlipAngle": 90,
    "ImageOrientationPatientDICOM": [
        1,
        0,
        0,
        0,
        1,
        0
    ],
    "ImageType": [
        "ORIGINAL",
        "PRIMARY",
        "M",
        "ND",
        "MOSAIC"
    ],
    "ImagingFrequency": 123.259,
    "InPlanePhaseEncodingDirectionDICOM": "COL",
    "MRAcquisitionType": "2D",
    "MagneticFieldStrength": 3,
    "Manufacturer": "Siemens",
    "ManufacturersModelName": "TrioTim",
    "Modality": "MR",
    "PartialFourier": 1,
    "PatientPosition": "HFS",
    "PercentPhaseFOV": 100,
    "PhaseEncodingDirection": "j-",
    "PhaseEncodingSteps": 64,
    "PhaseResolution": 1,
    "PixelBandwidth": 2056,
    "ProcedureStepDescription": "mri_brain",
    "ProtocolName": "bbl1_idemo2_210",
    "PulseSequenceDetails": "%SiemensSeq%_ep2d_bold",
    "ReceiveCoilName": "32Ch_Head",
    "ReconMatrixPE": 64,
    "RepetitionTime": 3,
    "SAR": 0.253013,
    "ScanOptions": "FS",
    "ScanningSequence": "EP",
    "SequenceName": "_epfid2d1_64",
    "SequenceVariant": "SK",
    "SeriesDescription": "bbl1_idemo2_210",
    "SeriesNumber": 9,
    "ShimSetting": [
        4067,
        -6772,
        -3778,
        1390,
        -75,
        -153,
        624,
        34
    ],
    "SliceThickness": 3,
    "SliceTiming": [
        1.5075,
        0,
        1.5725,
        0.0675,
        1.6375,
        0.1325,
        1.7025,
        0.1975,
        1.7675,
        0.2625,
        1.835,
        0.3275,
        1.9,
        0.3925,
        1.965,
        0.46,
        2.03,
        0.525,
        2.095,
        0.59,
        2.1625,
        0.655,
        2.2275,
        0.72,
        2.2925,
        0.7875,
        2.3575,
        0.8525,
        2.4225,
        0.9175,
        2.49,
        0.9825,
        2.555,
        1.0475,
        2.62,
        1.115,
        2.685,
        1.18,
        2.75,
        1.245,
        2.815,
        1.31,
        2.8825,
        1.375,
        2.9475,
        1.44
    ],
    "SoftwareVersions": "syngo_MR_B17",
    "SpacingBetweenSlices": 3,
    "StudyID": "7686972",
    "StudyInstanceUID": "1.2.840.113745.101000.1002000.40728.4508.30537163",
    "TaskName": "idemo",
    "TotalReadoutTime": 0.0346501,
    "TxRefAmp": 330.354,
    "Obliquity": "False",
    "VoxelSizeDim1": 3.0,
    "VoxelSizeDim3": 3.0,
    "VoxelSizeDim2": 3.0,
    "Dim1Size": 64,
    "Dim2Size": 64,
    "Dim3Size": 46,
    "NumVolumes": 115
}
```

Does it make sense to [demote that check to a warning](https://github.com/FCP-INDI/C-PAC/commit/642cd534479ffa0f4827953868433aad1284ca0e#diff-def5baf5367f68f8ae4ac0303bd348dfb80934103e45d6d31282148870cb74c4) or is the fix more complicated? Or am I wanting to do something impossible?

shnizzedy commented:
```Python
220916-11:02:54,891 nipype.workflow ERROR:
	 Building workflow failed
Traceback (most recent call last):
  File "/code/CPAC/pipeline/cpac_pipeline.py", line 444, in run_workflow
    subject_id, sub_dict, c, p_name, num_ants_cores
  File "/code/CPAC/pipeline/cpac_pipeline.py", line 1117, in build_workflow
    wf, rpool = initiate_rpool(wf, cfg, sub_dict)
  File "/code/CPAC/pipeline/engine.py", line 1948, in initiate_rpool
    part_id, ses_id)
  File "/code/CPAC/pipeline/engine.py", line 1526, in ingress_raw_func_data
    data_paths['creds_path'], ses_id)
  File "/code/CPAC/utils/datasource.py", line 487, in ingress_func_metadata
    node, out_file = rpool.get('diffphase-dwell')[
  File "/code/CPAC/pipeline/engine.py", line 276, in get
    raise LookupError(info_msg)
LookupError: 

[!] C-PAC says: None of the listed resources are in the resource pool:

  diffphase-dwell

Options:
- You can enable a node block earlier in the pipeline which produces these resources. Check the 'outputs:' field in a node block's documentation.
- You can directly provide this required data by pulling it from another BIDS directory using 'source_outputs_dir:' in the pipeline configuration, or by placing it directly in your C-PAC output directory.
- If you have done these, and you still get this message, please let us know through any of our support channels at: https://fcp-indi.github.io/
```

```tree
└── ses-PNC1
    ├── anat
    │   ├── sub-2286509730_ses-PNC1_acq-refaced_T1w.json
    │   └── sub-2286509730_ses-PNC1_acq-refaced_T1w.nii.gz
    ├── fmap
    │   ├── sub-2286509730_ses-PNC1_magnitude1.json
    │   ├── sub-2286509730_ses-PNC1_magnitude1.nii.gz
    │   ├── sub-2286509730_ses-PNC1_magnitude2.json
    │   ├── sub-2286509730_ses-PNC1_magnitude2.nii.gz
    │   ├── sub-2286509730_ses-PNC1_phase1.json
    │   ├── sub-2286509730_ses-PNC1_phase1.nii.gz
    │   ├── sub-2286509730_ses-PNC1_phase2.json
    │   └── sub-2286509730_ses-PNC1_phase2.nii.gz
    └── func
        ├── sub-2286509730_ses-PNC1_task-frac2back_acq-VARIANTNumVolumes_bold.json
        ├── sub-2286509730_ses-PNC1_task-frac2back_acq-VARIANTNumVolumes_bold.nii.gz
        ├── sub-2286509730_ses-PNC1_task-frac2back_acq-VARIANTNumVolumes_events.tsv
        ├── sub-2286509730_ses-PNC1_task-idemo_bold.json
        ├── sub-2286509730_ses-PNC1_task-idemo_bold.nii.gz
        └── sub-2286509730_ses-PNC1_task-idemo_events.tsv
```

Should this even be looking for `DwellTime`?

sgiavasis commented:
@shnizzedy I updated this to revert to the original DCAN workflow's functionality, and to use readout time only if dwell time does not exist.

Are there any changes to your previous comments now that we've addressed the dwell time issues since then?

shnizzedy commented:
> Are there any changes to your previous comments now that we've addressed the dwell time issues since then?

I think all my previous comments were due to my at-the-time confusion about when `DwellTime` is necessary. I was specifically concerned with phase diff fieldmaps, so I don't think my comments are actually relevant to this PR at all. :upside_down_face:

sgiavasis commented:
> > Are there any changes to your previous comments now that we've addressed the dwell time issues since then?
> 
> I think all my previous comments were due to my at-the-time confusion about when `DwellTime` is necessary. I was specifically concerned with phase diff fieldmaps, so I don't think my comments are actually relevant to this PR at all. 🙃

Okay just wanted to make sure!

shnizzedy commented on CPAC/distortion_correction/utils.py:
```diff
@@ -170,29 +170,44 @@ def phase_encode(unwarp_dir, phase_one, phase_two, dwell_time_one,
+    meta_data = [dwell_time_one, dwell_time_two,
+                 ro_time_one, ro_time_two]
+    if not any(meta_data):
+        raise Exception("\n[!] Blip-FSL-TOPUP workflow: neither "
+                        "TotalReadoutTime nor DwellTime is present in the "
+                        "epi field map meta-data.")
```
```suggestion
        raise KeyError("\n[!] Blip-FSL-TOPUP workflow: neither "
                       "TotalReadoutTime nor DwellTime is present in the "
                       "epi field map meta-data.")
```

shnizzedy commented on CPAC/distortion_correction/utils.py:
```diff
@@ -170,29 +170,44 @@ def phase_encode(unwarp_dir, phase_one, phase_two, dwell_time_one,
+        elif ro_time_one and ro_time_two:
+            ro_times = [f"-1 0 0 {ro_time_one}", f"1 0 0 {ro_time_two}"]
+        else:
+            raise Exception("[!] No dwell time or total readout time "
+                            "present for the acq-fMRI EPI field maps.")
```
I think
```suggestion
            raise KeyError("[!] No dwell time or total readout time "
                           "present for the acq-fMRI EPI field maps.")
```
would be appropriate, but maybe `LookupError` or `FileNotFoundError` if it's not necessarily the metadata key that's missing

shnizzedy commented on CPAC/distortion_correction/utils.py:
```diff
@@ -170,29 +170,44 @@ def phase_encode(unwarp_dir, phase_one, phase_two, dwell_time_one,
+        elif ro_time_one and ro_time_two:
+            ro_times = [f"0 -1 0 {ro_time_one}", f"0 1 0 {ro_time_two}"]
+        else:
+            raise Exception("[!] No dwell time or total readout time "
+                            "present for the acq-fMRI EPI field maps.")
```
```suggestion
            raise KeyError("[!] No dwell time or total readout time "
                           "present for the acq-fMRI EPI field maps.")
```

shnizzedy commented on CPAC/utils/build_data_config.py:
```diff
@@ -648,7 +648,7 @@ def get_BIDS_data_dct(bids_base_dir, file_list=None, anat_scan=None,
                                         "sub-{participant}/fmap/sub-{participant}"
                                         "*magnitud*.nii.gz")
-
+            
```
```suggestion

```

shnizzedy reviewed (approved):
Looks great to me!

I'd throw more specific exceptions, but I think it's fine to merge as-is.
